### DISCUSSION TITLE
Plan for .NET 6 and WinUI3

### DISCUSSION THREAD

**Discussion by robloo** [unoplatform/uno#6097] [General]
Both .NET 6 and WinUI3 are expected to be completed this November. However, while ~WinUI3 support should finally be coming to UWP this year~ [cancelled], there is no word on .NET6 and Native AOT updates for UWP. This creates an interesting problem that has been talked about for several years. It's time to bring Uno into the discussion directly though to understand where they see the project going. This affects long-term planning for current UWP app developers who MUST update to the latest versions of .NET with or without UWP if Microsoft continues to leave it behind.

**UWP vs. Desktop apps**

Microsoft is half-heartedly supporting UWP and is clearly pushing devs to use .NET 6 with WinUI3 and CsWinRT as a desktop app instead of the unifying UWP tech. However, UWP will likely continue to be supported for years to come at least with maintenance. This creates some options, and unfortunately, branching code bases.

 * **Option 1**: The Uno Platform continues to support UWP and WinUI3 side-by-side. This is not only resource draining but likely unnecessary if UWP never advances past .NET Core 2.x and C#7.3. Developers must use the latest tooling and are being left behind by several important libraries now.
 * **Option 2**: Uno Platform transitions to desktop-first but continues with mobile support. This means WinUI3 becomes dominant and the Windows.UI.Xaml namespace components are entirely removed. This is likely the best approach for everyone as maintaining 2 versions of the controls is getting ugly and a lot of redundant code and complex build transforms need to be stripped from Uno.

So it seems come November Uno Platform should actually ONLY support WinUI3 namespace controls. And ONLY support UWP heads that use WinUI3, .NET6 and CsWinRT. Of course Android/iOS apps will still be possible using Uno; however, they will need to be using Microsfot.UI.Xaml controls.

Thoughts?

**Comment by francoistanguay** [comment]
So far, we've been able to get away with namespace renames during CI between WUX and MUX so it isn't resource intensive. At this point, we're confident we can keep both until the story stabilizes but you'll definitely see more activty on the WInUI 3 side as it gains in traction.

**Reply by robloo** [reply]
I certainly understand the wait and see approach. Apps are quickly going to have to switch to WinUI/CsWinRT almost exclusively to get latest .NET so as long as you have a good solution there (and you do) I guess there isn't much need to change the current strategy. I was thinking its a great opportunity to clean up the code base and remove some gotcha's like adding spaces in namespaces (or different control implementations like NavigationView). However, when I started the discussion I was probably more in an App instead of framework/platform mindset.

**Comment by robloo** [comment]
As of yesterday Microsoft fired a pretty big shot into UWP and is starting to officially acknowledge it has no future. This is all part of the process Microsoft uses to retire technology and they are never very direct about it. You probably have much more insight into it than I do.

That said, what is the plan for WinRT in the Uno Platform now that UWP is officially a dead-end? The entirety of Uno seems to need to be restructured to support normal .NET6 apps with whatever API's that supports cross-platform (no more WinRT). Uno will need to essentially cut away most of WinRT and re-use only those components that are in WinUI 3 (drag-and-drop, etc.). The XAML UI layer thankfully can stay mostly as-is; this just affects the app model and device integration.

**Reply by francoistanguay** [reply]
UWP means many things. WinRT apis aren't going anywhere. The container/isolation aspects of UWP will be replaced and folded in Windows App SDK, the WinUI UI technology will live on its own. It doesnt change anything for Uno. It was WinUIEverywhere (with WinRT Non-UI api support) and will continue to be. 

Happy to discuss further if I missed something.

**Reply by robloo** [reply]
It's certainly true that UWP is now being broken up into an assortment of technologies and each will 'live' various places going forward (mostly under Windows App SDK). However, fundamentally, I don't see what use WinRT is without UWP and I expect no one will write new apps using WinRT going forward. WinRT is maintained but on life support just the same as UWP was (no new features). So at this point I disagree with "WinRT apis aren't going anywhere" in practice (attrition over time will kill their usage as Microsoft likely intends).

The question I struggle with is what does this mean for Uno?

With Uno, WinRT cross-platform ready API's is a pretty big feature. I guess the disconnect we have is I see a clear dead-end using them. I suspect this is fundamentally going to cause a course correction with Uno and am trying to understand what that will look like. Microsoft is going two directions:

 1. WinUI team is moving back to promoting and supporting Win32 API's or .NET+WinUI as a desktop app only
 2. .NET team is moving to MAUI and promoting the various cross-platform APIs that includes (of course .NET itself has many cross-platform APIs as well)

There is no longer one direction to follow and this seems like a clear branch in the road. As Uno was following path 1, and path 1 is essentially dropping multi-device support and WinRT, you have to decide what your API surface area is going to look like. What makes sense for app developers now?

What makes sense is basically .NET+WinUI without WinRT. However, what about all the API's to access device properties that were formerly in WinRT but aren't in .NET? Equivalents usually exist now in MAUI. Uno almost seems to need to switch to adopting the MAUI device API's going forward and build on top MAUI desktop for even Windows.

So again, I see lots a lot of divergence coming up. It seems Uno should now actually be moving to the MAUI app model and cross-platform device API's. Fundamentally you support .NET and C# apps -- not native Win32 apps. Following the WinUI team is a dead-end for anything other than Windows desktop. Who knows what will happen to WinRT in even a few years from now (I expect nothing good).

**Comment by francoistanguay** [comment]
Don't get me wrong, you're right on many aspects and we're doing our best to plan according to how the underlying technologies are evolving. I could agree that continuing to support WinRT APIs offer more choice and therefore can lead to confusion. On the flip side, you can use .NET-only APIs and could consume Microsoft.MAUI.Essentials just like you can consume Xamarin.Essentials today.

But you're also right that we're not in the business of trying to bring Native Win32 APIs x-plat either.

If/When it becomes clear that WinRT is no longer, we could reassess but this is NOT our understanding as-of today. 

Fun times...

**Reply by robloo** [reply]
WinUI3 has had a lot of restructuring behind the scenes but as far as I understand it, it is still using WinRT and will stay that way. So you're absolutely right WinRT isn't going to be removed from Windows anytime soon. I just don't see how apps will (or even should) be using it going forward. There also is a strong shift where .NET seems to be pulling away from WinUI focusing on MAUI so that complicates things a bit as mentioned before.

The problem is not the UI-layer though, it's the cross-platform device API's. In the end, I think the compatiblity story with existing code will need to change. Uno will need to create it's own path forward on Windows desktop.

This could play-out a few different ways:
 1. Keep using WinRT. This is likely not ideal for app developers a few years in the future. Nobody will be using WinRT for anything other than Uno itself. UWP apps will be re-written using other tech.
 1. As MAUI becomes the de-facto cross-platform API and UI layer for .NET, Uno could build on top of it instead. This would give access to the newest device API's (and the ones most integrated with .NET). Uno is still by far the more powerful UI-layer and I don't expect there to be much competition from that aspect, you would just be using their app model and device API's instead of WinRT.
 1. Even more unorthodox: A partnership with Avalonia and Uno. Basically, Uno Platform runs on top of the Avalonia cross-platform Skia-powered backend. That would provide a powerful option completely separate from MAUI. This would involve standardizing on a brand new set of API's though so the argument for compatiblity goes away completely.

Personally, I'm going to be moving most WinRT API calls to an abstraction layer with the intention of removing them in the (near?) future. I should emphasize that in apps WinRT usage seems to be minimal. The vast majority of code is .NET and the WinUI stack. So one way or another it shouldn't be a huge tear-up for app developers. It's just defining what API's will take the place of WinRT I'm trying to understand. It's not doom and gloom 😄 

---

Finally, to vent, what's most sad about all this is WinUI3 was the fix for UWP. Pulling out of the OS allowed fixing a lot of the bugs and instabilities that prevented uptake in the beginning. A few years of open source development would have made UWP rock-solid I think. It is still a beautiful API surface in my opinion but as Microsoft tends to do, investment stopped before it started paying off.

UWP with Uno is still a wonderful technology that is extremely powerful for cross-platform development. Combined with C# I don't think it has an equal. WinRT is only a subset of it's usefulness!

**Reply by francoistanguay** [reply]
Couple of thoughts:

It's totally fine that you go down the route of creating your own abstraction layer. You could argue that MAUI.Essentials is already an abstraction and could be relied upon directly although they still don't have interfaces (!!!).

Another way to look at it is that the WinRT APi contract is pretty rich and could be your abstraction. That's what Uno is bringing to other platforms and if WinRT was to be "replaced" by something else somehow, someone (or Uno) could offer a bridge from the rich WinRT API to whatever happens in the future.

Uno's job today is to bring WinUI everywhere (and as much from Windows App SDK that makes sense in a cross platform scenario, if anything), therefore porting WinRT APIs still make sense today.

As for the unrelated topic of Uno sitting on top of Avalonia, it doesn't make sense to us. First, we already have a Skia backend if/when needed. Second, we took an explicit direction of aligning with Microsoft's ecosystem so third party components can easily be reused.

**Comment by marwalsch** [comment]
I noticed that even in the .NET 6 Uno templates, the current `Uno.UI` namespaces are still targeted rather than the "subsequent" `Uno.WinUI` packages. Is there any particular reason WinUI 3 is not becoming the default for entirely new projects created after the release of .NET 6? From my state of knowledge UWP Xaml is not currently compatible with .NET 5 and onward.
